drivers/volume/portworx/portworx.go

```diff
@@ -923,6 +923,140 @@ func (p *portworx) SnapshotDelete(snapDataSrc *crdv1.VolumeSnapshotDataSource, _
+		// GetSnapshot Details
+		snapshot, err := k8s.Instance().GetSnapshot(snapName, snapNamespace)
+		if err != nil {
+			return fmt.Errorf("failed to retrieve VolumeSnapshot %s: %v",
+				snapName, err)
+		}
+		logrus.Infof("Snapshot found %v", snapshot.Metadata.Name)
+		if len(snapshot.Spec.SnapshotDataName) == 0 {
+			return fmt.Errorf("VolumeSnapshot %s is not bound to any VolumeSnapshotData",
+				snapName)
+		}
+
+		snapshotData, err := k8s.Instance().GetSnapshotData(snapshot.Spec.SnapshotDataName)
+		if err != nil {
+			return fmt.Errorf("failed to retrieve VolumeSnapshotData %s: %v",
+				snapshot.Spec.SnapshotDataName, err)
+		}
+		// restore to snapshot PVC
+		volID, err := getVolumeIDFromPVC(
+			snapshot.Spec.PersistentVolumeClaimName,
+			snapshot.Metadata.Namespace)
+		if err != nil {
+			return fmt.Errorf("failed to get volume ID for snapshot %v", err)
+		}
+
+		snapID := snapshotData.Spec.PortworxSnapshot.SnapshotID
+		restoreVolumes[snapID] = volID
+	}
+	logrus.Debugf("Restore volume map %v", restoreVolumes)
+
+	switch snapRestore.Spec.SourceType {
+	case "", string(crdv1.PortworxSnapshotTypeLocal):
+		return p.localSnapshotRestore(restoreVolumes, params)
+	case string(crdv1.PortworxSnapshotTypeCloud):
+		return p.cloudSnapshotRestore(restoreVolumes, params)
+	default:
+		return fmt.Errorf("invalid SourceType for snapshot(local/cloud)")
+	}
+}
+
+func getRestoreVolumeMap(snapName, snapNamespace string) (map[string]string, error) {
+	volumes := make(map[string]string)
+	groupSnaps, err := k8s.Instance().GetSnapshotsForGroupSnapshot(snapName, snapNamespace)
+	if err != nil {
+		logrus.Errorf("Unable to get group snapshot details %v", err)
+		return volumes, err
+	}
+
+	for _, snapForGroupsnap := range groupSnaps {
+		snapID := string(snapForGroupsnap.Metadata.UID)
+		logrus.Debugf("Getting volume ID for pvc %v", snapForGroupsnap.Spec.PersistentVolumeClaimName)
+		volID, err := getVolumeIDFromPVC(
+			snapForGroupsnap.Spec.PersistentVolumeClaimName,
+			snapForGroupsnap.Metadata.Namespace)
+		if err != nil {
+			return volumes, fmt.Errorf("failed to get volume ID for snapshot %v", err)
+		}
+		volumes[snapID] = volID
+	}
+	return volumes, nil
+}
+
+func (p *portworx) localSnapshotRestore(
+	restoreVolumes map[string]string,
+	params map[string]string,
+) error {
+	// Get volume client from user context
+	volDriver, err := p.getUserVolDriver(params)
+	if err != nil {
+		return err
+	}
+	volumeInfo := []*stork_crd.RestoreVolumeInfo{}
+	for snapID, volID := range restoreVolumes {
+		logrus.Infof("Restoring volume %v with local snapshot %v", volID, snapID)
+		err = volDriver.Restore(volID, snapID)
+		if err != nil {
+			logrus.Errorf("Unable to restore volume %v with ID %v, err %v",
+				volID, snapID, err)
+			return err
+		}
+		logrus.Infof("Completed restore for volume %v with Snapshotshot %v", volID, snapID)
+		// TODO: should we fail if one snap restore fails or retiries for failed
```

Comment:
We shouldn't retry here. We should return an error from the driver to the controller and the controller will retry in the next reconcile loop